include/broker/format/json.hh

```diff
@@ -0,0 +1,324 @@
+
+/// Appends a string to the output iterator.
+template <class OutIter>
+OutIter append(std::string_view str, OutIter out) {
+  return std::copy(str.begin(), str.end(), out);
+}
+
+/// Tag type for selecting the `quoted` overload of `append_encoded`.
+struct quoted {
+  std::string_view str;
+};
+
+/// Appends a quoted string to the output iterator. Special characters are
+/// escaped.
+template <class OutIter>
+OutIter append(quoted value, OutIter out) {
+  using namespace std::literals;
+  *out++ = '"';
+  for (auto c : value.str) {
+    switch (c) {
+      default:
+        *out++ = c;
+        break;
+      case '\\':
+        *out++ = '\\';
+        *out++ = '\\';
+        break;
+      case '\b':
+        *out++ = '\\';
+        *out++ = 'b';
+        break;
+      case '\f':
+        *out++ = '\\';
+        *out++ = 'f';
+        break;
+      case '\n':
+        *out++ = '\\';
+        *out++ = 'n';
+        break;
+      case '\r':
+        *out++ = '\\';
+        *out++ = 'r';
+        break;
+      case '\t':
+        *out++ = '\\';
+        *out++ = 't';
+        break;
+      case '\v':
+        *out++ = '\\';
+        *out++ = 'v';
+        break;
+      case '"':
+        *out++ = '\\';
+        *out++ = '"';
+        break;
+    }
+  }
+  *out++ = '"';
+  return out;
+}
+
+/// Appends a field (key-value pair) to the output iterator.
+template <class OutIter>
+OutIter append_field(std::string_view key, std::string_view val, OutIter out) {
+  out = append(quoted{key}, out);
+  *out++ = ':';
+  return append(quoted{val}, out);
+}
+
+/// Appends an encoded value to the output iterator.
+/// @param value_type The type name of the value.
+/// @param value The value. Must provide an overload for `append()`.
+/// @param out The output iterator.
+template <class Policy = render_object, class Appendable, class OutIter>
+OutIter append_encoded(std::string_view value_type, Appendable value,
+                       OutIter out) {
+  using namespace std::literals;
+  if constexpr (std::is_same_v<Policy, render_object>)
+    *out++ = '{';
+  out = append(R"_("@data-type":")_", out);
+  out = append(value_type, out);
+  out = append(R"_(","data":)_", out);
+  out = append(value, out);
+  if constexpr (std::is_same_v<Policy, render_object>)
+    *out++ = '}';
+  return out;
+}
+
+/// Render the `nil` value to `out`.
+template <class Policy = render_object, class OutIter>
+OutIter encode(none, OutIter out) {
+  return append_encoded<Policy>("none", "{}", out);
+}
+
+/// Renders `value` to `out` as `T` for `true` and `F` for `false`.
+template <class Policy = render_object, class T, class OutIter>
+std::enable_if_t<std::is_integral_v<T>, OutIter> encode(T value, OutIter out) {
+  using namespace std::literals;
+  if constexpr (std::is_same_v<T, bool>) {
+    return append_encoded<Policy>("boolean", value ? "true"sv : "false"sv, out);
+  } else if constexpr (std::is_same_v<T, count>) {
+    // An integer can at most have 20 digits (UINT64_MAX).
```

Comment:
Vanilla JavaScript cannot correctly parse numbers so large. Nothing to do, but that would probably be a topic if we had more browsers interact with Zeek.

This even made it into the RFC :-)

>  https://datatracker.ietf.org/doc/html/rfc7159#section-6
>
>   Note that when such software is used, numbers that are integers and
>   are in the range [-(2**53)+1, (2**53)-1] are interoperable in the
>   sense that implementations will agree exactly on their numeric
>   values.

Reply:
Probably because JavaScript will store numbers as 64-bit floating point numbers? Probably not much we could do about that. Except sending integers as strings... 🤔

Reply:
Yeah, I don't think we should worry too much. If someone actually wants to work with numbers that large, they'll need to use BigInt anyhow and there's `json-bigint` that can parse this properly. 

---

Where I saw this first:

> In newer versions of the API, all large integer values are represented as strings by default.

https://developer.twitter.com/en/docs/twitter-ids